gitpod-io/gitpod
# JB gateway plugin

## Description


This PR refactors JB integration to leverage SSH Gateway and enable client-to-server flow from JB Gateway, particularly:
- fixing JB backend plugin
    - remove dependencies, since JB backend already has Kotlin dependencies, and adding other version leads to runtime incompatibilities
    - fix memory leak of grpc message channels
    - use Kotlin koroutines instead of dedicated thread for heartbeating
    - implement web socket reconnection handling for Gitpod server connection
    - pin java target to v.11 in all projects
- initial Gitpod plugin for JB Gateway that provides:
    - connection over SSH gateway for `jetbrains-gateway://connect?gitpodHost=<gitpodHost>&workspaceId=<workspaceId>` links
    - initial implementation of Gitpod UI in JB Gateway, i.e. welcome screen and recent projects view
- switch JB images to use JB Gateway with Gitpod plugin in order to establish connection over SSH gateway not CWM links
- add JB Plugins as client types for server tracing
- resume hearbeating if window unload was cancelled in supervisor frontend on redirection to JB Gateway

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/5644, #7524

## How to test


From Gitpod:
- Go to preferences and switch to IntelliJ: https://ak-jb-gateway-plugin.staging.gitpod-dev.com/preferences
- Start a workspace and check that you can connect.

From Gateway:
- If you did above and everything was clear by now you should have JB Gateway installed.
- `Preferences...` -> `Tools` -> `Gitpod` and point Gitpod Host to `ak-jb-gateway-plugin.staging.gitpod-dev.com`
- Try to use Gateway to start workspaces from there using Gitpod UI.

## Release Notes

```release-note
- Refactor JB integration to connect over SSH instead of CWM links.
- Provide Gitpod integration in JB Gateway.
```

## Documentation
